cfgrammar/src/lib/yacc/parser.rs

```diff
@@ -224,109 +234,55 @@ impl YaccParser {
-            let mut tmp = vec![*orig_span];
-            tmp.extend(spans);
-            return Err(vec![YaccGrammarError {
-                kind: YaccGrammarErrorKind::DuplicatePrecedence,
-                spans: tmp,
-            }]);
-        }
-        if let Some((orig_span, spans)) = self.duplicate_implicit_token_spans.iter().next() {
-            let mut tmp = vec![*orig_span];
-            tmp.extend(spans);
-            return Err(vec![YaccGrammarError {
-                kind: YaccGrammarErrorKind::DuplicateImplicitTokensDeclaration,
-                spans: tmp,
-            }]);
-        }
-        if let Some((orig_span, spans)) = &self.duplicate_expect_declarations {
-            let mut tmp = vec![*orig_span];
-            tmp.extend(spans);
-            return Err(vec![YaccGrammarError {
-                kind: YaccGrammarErrorKind::DuplicateExpectDeclaration,
-                spans: tmp,
-            }]);
-        }
-        if let Some((orig_span, spans)) = &self.duplicate_expectrr_declarations {
-            let mut tmp = vec![*orig_span];
-            tmp.extend(spans);
-            return Err(vec![YaccGrammarError {
-                kind: YaccGrammarErrorKind::DuplicateExpectRRDeclaration,
-                spans: tmp,
-            }]);
-        }
-        if let Some((orig_span, spans)) = &self.duplicate_start_declarations {
-            let mut tmp = vec![*orig_span];
-            tmp.extend(spans);
-            return Err(vec![YaccGrammarError {
-                kind: YaccGrammarErrorKind::DuplicateStartDeclaration,
-                spans: tmp,
-            }]);
-        }
-        if let Some((orig_span, spans)) = self.duplicate_epp_declarations.iter().next() {
-            let mut tmp = vec![*orig_span];
-            tmp.extend(spans);
-            return Err(vec![YaccGrammarError {
-                kind: YaccGrammarErrorKind::DuplicateEPP,
-                spans: tmp,
-            }]);
-        }
-        if let Some(spans) = &self.duplicate_actiontype_declarations {
-            let orig_span = *self
-                .global_actiontype
-                .as_ref()
-                .map(|(_, span)| span)
-                .unwrap();
-            let mut tmp = vec![orig_span];
-            tmp.extend(spans);
-            return Err(vec![YaccGrammarError {
-                kind: YaccGrammarErrorKind::DuplicateActiontypeDeclaration,
-                spans: tmp,
-            }]);
-        }
-        i = self.parse_rules(i?).map_err(|e| vec![e]);
-        if let Some((orig_span, spans)) = self.duplicate_rule_spans.iter().next() {
-            let mut tmp = vec![*orig_span];
-            tmp.extend(spans);
-            return Err(vec![YaccGrammarError {
-                kind: YaccGrammarErrorKind::DuplicateRule,
-                spans: tmp,
-            }]);
+        let mut result = self.parse_declarations(0, &mut errs);
+        result = self.parse_rules(
```

Comment:
I wonder if we can squeeze this down to something like `self.parse_rules(result.map_err(|e| errs.push(e); Err(errs))?, &mut errs)`? The borrow checker might not like it though.

Reply:
Yeah, that won't work exactly because of the borrow checker, and most of the other combinators won't allow a `return` from within, so I've been perplexed as to whether there is a nicer way.

Reply:
Let's stick with it as-is. We can fix it in a later PR if we think of something clever.